lib/vector-core/src/enrichment_table/enrichment_tables.rs

```diff
@@ -0,0 +1,286 @@
+use super::EnrichmentTable;
+use arc_swap::ArcSwap;
+#[cfg(feature = "vrl")]
+use std::collections::BTreeMap;
+use std::sync::Arc;
+use std::{collections::HashMap, sync::Mutex};
+
+/// The Enrichment Tables manages the collection of `EnrichmentTable`s loaded into Vector.
+/// Enrichment Tables go through two stages.
+///
+/// ## 1. Writing
+///
+/// The  tables are loaded. There are two elements that need loading. The first is the actual data.
+/// This is loaded at config load time, the actual loading is performed by the implementation of
+/// the `EnrichmentTable` trait. Next, the tables are passed through Vectors `Transform` components,
+/// particularly the `Remap` transform. These Transforms are able to determine which fields we will
+/// want to lookup whilst Vector is running. They can notify the tables of these fields no that the
+/// data can be indexed.
+///
+/// During this phase, the data needs to be mutated and shared around a number of potential
+/// different components. Also, performance is not absolutely critical at this phase. Thus, the
+/// data is stored within a `Mutex` - stored in the `loading` field of this struct.
+///
+/// ## 2. Reading
+///
+/// Once all the data has been loaded we can move to the next stage. This is signified by calling
+/// the `finish_load` method. At this point all the data is swapped out of the `Mutex` in the loading
+/// field into the `ArcSwap` of the `tables` field. `ArcSwap` provides lock-free read-only access
+/// to the data. From this point on we have fast, efficient read-only access and can no longer add
+/// indexes or otherwise mutate the data .
```

Comment:
So, we're dealing with a state machine that transitions from the `startup` state to the `available` (?) state, can't revert back, and has different interfaces depending on the state. It might be worth making this more explicit by using an actual enum or any library that provides some convenience around state management.

Reply:
The issue with making this an enum is that when changing the state, we would need to mutate *self. Since by this point there are many clones of `EnrichmentTables` around, when mutating *self, you wouldn't be mutating the underlying data of all the clones.

So you could perhaps wrap an enum in a struct that is passed around and mutate the enum, however for all the clones of the struct to share the underlying enum you would need to wrap it in an `Arc<Mutex>` or similar, which would then defeat the whole purpose.

Reply:
Each Transform is passed this structure in the `build`. The remap transform will compile the VRL source, during with point it will do the loading. It then stores the structure so that it can use it for reading during run time.

I could update `TransformContext` to be 

```
pub struct TransformContext {
    pub globals: GlobalOptions,
    pub enrichment_tables_read: EnrichmentTablesRead,
    pub enrichment_tables_write: EnrichmentTablesWrite,
}
```

Writing to the Writer would still cause an error until the end of the load stage, so essentially you would be passing around an object that is invalid until a certain point. But, I could make the writer so it isn't `Clone`, thus preventing the Transform from storing it, which would force it to do all the writing at load time.

Another possibility could be to add a `get_reader(&self) -> EnrichmentTablesRead` function to the Writer. The Writer would store both the `Mutex` and the `ArcWrap`, but would only provide access to the `Mutex`. You would have to call `get_reader` to get the object that provides the read functionality. The read object would also be `Clone` so you would need to call that in order to store it.

I think I quite like that last idea.

Reply:
> Is this still something you're looking into changing,

Yes, I'm hoping to push this in a minute.

Reply:
I have pushed this now, so there is a separate `Tables` and `TableSearch` struct. One unexpected bonus of splitting this struct up has been that I can get rid of the Mutex, so now both reads and writes are lock free!

Reply:
Unfortunately, I spoke too soon. Removing the Mutex makes it really complicated passing a reference to the tables through the Vrl compiler. So it seems the Mutex will return in the next PR.